feat(sdk): react x wing integration (#3958)
winglang/wing: libs/wingsdk/src/ex/react-app.md

---

> title: React Website

Reviewer: Name consistency - shouldn't this match the .md file name? I'd expect both to be React App.
```suggestion
title: React App
```

> id: react-website

Reviewer: ```suggestion
id: react-app
```

description: A built-in resource for creating deployable websites using the React framework.
keywords: [Website, React, deployment, build]
sidebar_position: 1
---

The `ex.ReactApp` resource represents a website, built using React, that can be both hosted in the cloud or to run on a development hot-reload server in the sim target.

## Usage

### Initialization

Using the default arguments:

```ts
bring ex;
bring util;

let website = new ex.ReactApp(projectPath: "./client");
```

or customizing them:

```ts
bring ex;
bring util;

let website = new ex.ReactApp(
  projectPath: "./client",

>   isDevRun: false // `true`` by default. will run the start command if true, and the build command if not

Reviewer: ```suggestion
  isDevRun: false // `true` by default. Will run the start command if true, and the build command if not
```

  buildDir: "/dist" // default is "/build"
  startCommand: "pnpm start" // default is "npm start"
  buildCommand: "pnpm build" // default is "npm build"
  localPort: 4000 // default is 3001
 );
```

When `ReactApp` is compiled to the `sim` target, by default it runs the start command (default: `npm start`) inside `projectPath` to serve your app in development mode on a local port.

If the `CI` environment variable is set OR if `ReactApp` is compiled to any other target, it will run the build command (default: `npm build`) inside of the `projectPath` to build the React app for production to `buildDir` and serve the app.

> 

Reviewer 2: It doesn't look like this was implemented-

If we're keeping `isDevRun` then let's update the docs to make sure it's clear that the option only has an effect on the `sim` target, and that the `buildCommand` is always used for building the react app on other targets.

---
title: React Website
id: react-website
description: A built-in resource for creating deployable websites using the React framework.
keywords: [Website, React, deployment, build]
sidebar_position: 1
---

The `ex.ReactApp` resource represents a website, built using React, that can be both hosted in the cloud or to run on a development hot-reload server in the sim target.

## Usage

### Initialization

Using the default arguments:

```ts
bring ex;
bring util;

let website = new ex.ReactApp(projectPath: "./client");
```

or customizing them:

```ts
bring ex;
bring util;

let website = new ex.ReactApp(
  projectPath: "./client",
  isDevRun: util.tryEnv("ENV") == "dev" // `true`` by default. will run the start command if true, and the build command if not
  buildDir: "/dist" // default is "/build"
  startCommand: "pnpm start" // default is "npm start"
  buildCommand: "pnpm build" // default is "npm build"
  hostProps: {} // properties that apply to the react app host, which is a `cloud.Website` resource

>   localPort: "4000" // default is 3001

Reviewer 2: A few thoughts/questions:
1. Should the default be a random local port? If you have more than one react app, they will both try to serve to 3001 and it could error.
2. It smells a little weird that this is customizable for ReactApp but not for Website.
3. If this option is relevant for the `sim` target I'm not sure we should have it on the public, target-agnostic API - usually the public API only has options that can be configured for all targets. I guess this is related to https://github.com/winglang/wing/issues/2271. All that said I'm ok keeping `localPort` for now and revisiting this when we decide on a solution for `Api`

Author: it's only customizable for running react on dev mode (when building it it just get the port of the hosting website), but I agree with the rest.
it shouldn't be a random port since it will make it impossible to develop in dev mode- it will close the react dev server every time the wing file updates/ running `wing it` which breaks the hot reloading (and a bad experience for the developer too)

 );
```

- When calling `wing it` with `isDevRun` set to `true`, Wing will start both the simulator and React's development server.
- When calling `wing it` or `wing compile` for the `sim` target with `isDevRun` set to `false`, Wing will build React and serve it via the [Website resource](../01-cloud/website.md).
- When calling `wing compile` for any other target, Wing will build React locally and generate output files, preparing it for deployment to the cloud using the [Website resource](../01-cloud/website.md).

> ### Using wing variables within react code

Reviewer 2: ```suggestion
### Using Wing variables within react code
```

> #### using wing variables within react code

Reviewer 2: ```suggestion
### Using wing variables within react code
```

> `ex.ReactApp` allows you to pass preflight arguments from wing to the React app using `addEnvironment` method:

Reviewer 2: ```suggestion
`ex.ReactApp` allows you to pass preflight string values from Wing to the React app using `addEnvironment` method:
```

> `ex.ReactApp` allows you to pass preflight arguments from wing to the React app using `addEnvironmentVariable` method:

Reviewer: > @staycoolcall911 I don't have a strong preference between these names (maybe @eladb has an opinion?), but the name should match the corresponding method in cloud.Function since it has the same functionality.

@Chriscbr - thanks for the feedback, I wasn't aware of the `cloud.Function` method. In that case, let's keep it consistent. Sorry for the back and forth @tsuf239... let's revert to use `addEnvironment`
